demos/widgets/lv_demo_widgets.py

```diff
@@ -1,572 +1,1139 @@
+    label.set_text(str(val) + "Mbps")
+
+def ta_event_cb(e,kb,tv) :
+
+    code = e.get_code()
+    ta = e.get_target()
+    # lv_obj_t * kb = lv_event_get_user_data(e);
+    
+    indev = lv.indev_get_act()
+    if code == lv.EVENT.FOCUSED :
+        if indev.get_type() != lv.INDEV_TYPE.KEYPAD :
+            kb.set_textarea(ta)
+            kb.set_style_max_height(LV_HOR_RES * 2 // 3, 0)
+            tv.update_layout()   # Be sure the sizes are recalculated
+            tv.set_height(LV_VER_RES - kb.get_height())
+            kb.clear_flag(lv.obj.FLAG.HIDDEN)
+            ta.scroll_to_view_recursive(lv.ANIM.OFF)
+ 
+    elif code == lv.EVENT.DEFOCUSED :
+        kb.set_textarea(None)
+        tv.set_height(LV_VER_RES)
+        kb.add_flag(lv.obj.FLAG.HIDDEN)
+        indev.reset(None)
+
+    elif code == lv.EVENT.READY or code == lv.EVENT.CANCEL:
+        tv.set_height(LV_VER_RES)
+        kb.add_flag(lv.obj.FLAG.HIDDEN)
+        ta.clear_state(lv.STATE.FOCUSED)
+        indev.reset(None)    # To forget the last clicked object to make it focusable again
+
+def birthday_event_cb(e):
+    global calendar
+    code = e.get_code()
+    ta = e.get_target()
+
+    if code == lv.EVENT.FOCUSED :
+        indev = lv.indev_get_act()
+        if indev.get_type() == lv.INDEV_TYPE.POINTER:
+            if calendar == None:
+                top_layer = lv.layer_top()
+                top_layer.add_flag(lv.obj.FLAG.CLICKABLE)
+                calendar = lv.calendar(top_layer)
+                top_layer.set_style_bg_opa(lv.OPA._50, 0)
+                top_layer.set_style_bg_color(lv.palette_main(lv.PALETTE.GREY), 0)
+                if disp_size == DISP_SMALL:
+                    calendar.set_size(180, 200)
+                elif disp_size == DISP_MEDIUM:
+                    calendar.set_size(200, 220)
+                else :
+                    calendar.set_size(300, 330) 
+                calendar.set_showed_date(1990, 01)
+                calendar.align(lv.ALIGN.CENTER, 0, 30)
+                # password.add_event_cb(lambda e: ta_event_cb(e,kb,tv),lv.EVENT.ALL,None)
+              
+                calendar.add_event_cb(lambda e: calendar_event_cb(e,ta), lv.EVENT.ALL, None)
+
+                header = lv.calendar_header_dropdown(calendar)
+
+def calendar_event_cb(e,ta) :
+    global calendar
+    code = e.get_code();
+    obj = lv.calendar.__cast__(e.get_current_target())
+    if code == lv.EVENT.VALUE_CHANGED:
+        d = lv.calendar_date_t()
+        obj.get_pressed_date(d)
+        ta.set_text("{:02d}.{:02d}.{:d}".format(d.day,d.month,d.year))
+
+        calendar.delete()
+        calendar = None
+        lv.layer_top().clear_flag(lv.obj.FLAG.CLICKABLE)
+        lv.layer_top().set_style_bg_opa(lv.OPA.TRANSP, 0)
+   
+t1 = tv.add_tab("Profile")
+t2 = tv.add_tab("Analytics")
+t3 = tv.add_tab("Shop")
+
+profile_create(t1)
+analytics_create(t2)
+shop_create(t3)
+
+color_changer_create(tv)
+
+while True:
```

Comment:
I think it would be a good idea to add this demo to the automated testing.  
Today every PR or commit goes through a series of tests (GitHub action CI workflow).   
One of these tests involve building the Micropython unix port **and running all the examples**.  
The advantage is that if some change breaks the example code, we know about this immediately.

Your demo can also run automatically, but not with this `while True:` loop.  
So I suggest removing it.  

We can document that on the unix port this needs to be run with `-i` command line option to prevent the window from closing.

Here is how to run your demo with the test automation:

```bash
ports/unix/micropython-dev lib/lv_bindings/tests/run_test.py lib/lv_bindings/lvgl/demos/widgets/lv_demo_widgets.py
```

When it's ready, I will add it to [the script](https://github.com/lvgl/lv_binding_micropython/blob/6431ec4b1c66c92880a533020eb4802ea43694b0/tests/run.sh#L9) so that would run automatically.

Reply:
loop removed